examples/authentication/client/src/main.rs

```diff
@@ -0,0 +1,194 @@
+
+use futures::channel::oneshot::{self, Sender};
+use futures_util::future;
+use hyper::{service::Service, Body, Request, Response, Server, StatusCode};
+use log::{info, warn};
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+};
+use structopt::StructOpt;
+use url::{form_urlencoded, Url};
+
+#[derive(StructOpt, Clone)]
+#[structopt(about = "OpenID Connect Client Example.")]
+pub struct Opt {
+    #[structopt(
+        long,
+        help = "Address of the Oak application to connect to.",
+        default_value = "127.0.0.1:8080"
+    )]
+    address: String,
+    #[structopt(
+        long,
+        help = "Address to listen on for the OAuth redirect.",
+        default_value = "127.0.0.1:8089"
+    )]
+    redirect_address: String,
+    #[structopt(long, help = "Path to the PEM-encoded CA root certificate.")]
+    ca_cert: Option<String>,
+    // TODO(#886): Retrieve client_id from server, rather than flag.
+    #[structopt(long, help = "The OAuth client ID.")]
+    client_id: String,
+}
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    env_logger::init();
+    let opt = Opt::from_args();
+    let redirect_address = opt.redirect_address.parse()?;
+    info!("Listening for redirect on {}", &redirect_address);
+
+    info!("Opening Auth request in Browser");
+    let request_uri = get_request(opt);
+    info!("URI: {}", &request_uri);
+    let (sender, receiver) = oneshot::channel();
+    let result_sender = Arc::new(Mutex::new(ResultSender {
+        sender: Some(sender),
+        result: None,
+    }));
+    let maker = MakeHandler {
+        result_sender: result_sender.clone(),
+    };
+
+    let mut runner = tokio::runtime::Runtime::new().unwrap();
+    let task = runner.spawn(async move {
+        Server::bind(&redirect_address)
+            .serve(maker)
+            .with_graceful_shutdown(async move {
+                receiver.await.unwrap();
+            })
+            .await
+    });
+    open::that(request_uri)?;
+    runner.block_on(task)??;
+
+    let result = result_sender.lock().unwrap();
+    let code = result.result.as_ref().unwrap().as_ref().ok().unwrap();
+    info!("Received code: {}", code);
+    Ok(())
+}
+
+fn get_request(opt: Opt) -> String {
+    // TODO(#886): Retrieve endpoint from server.
+    let mut auth_endpoint = Url::parse("https://accounts.google.com/o/oauth2/v2/auth").unwrap();
+    // TODO(#886): Consider retrieving scope from server.
+    let scope = "openid email";
+    let redirect_uri = format!("http://{}", opt.redirect_address);
+    // TODO(#886): Retrieve state and code challenge from server and add to request.
+    auth_endpoint
+        .query_pairs_mut()
+        .append_pair("scope", scope)
+        .append_pair("response_type", "code")
+        .append_pair("redirect_uri", &redirect_uri)
+        .append_pair("client_id", &opt.client_id);
+    auth_endpoint.to_string()
+}
+
+struct ResultSender {
+    sender: Option<Sender<()>>,
```

Comment:
Does `Sender<Result<String, String>>` not work here, instead of the two separate fields? In fact if that works, we don't really need the struct at all.

Reply:
I started that way, but it had its own complications. The `Receiver` gets the result within the `async` section inside `with_graceful_shutdown`. This code runs in a different thread, so another cross-thread communication mechanism is needed to get the results from the `Receiver` back to the main thread after the server is shut down. It could in principle be done with two oneshot channels as well, but I am not sure that is any clearer.

Also, oneshot `Sender`s and `Receiver`s cannot be copied, so it would need something like Arc<Mutex<Option<Result<String, String>>>>.

Reply:
Ah I see, `Sender` and `Receiver` work with futures, not with threads, so probably not a great fit here. How about https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html instead? That should be nicer to use across threads.

Reply:
Using `mpsc` solves one problem, in that it removes the need for `Arc` and `Mutex`, but there is still a need for a second channel. There are two signals needed. One to give the results to the main thread (which uses `result` on the struct) and one to notify Hyper to shut down the web server (which uses `sender` on the struct).

Reply:
I have now refactored the code to use `mspc` to notify `main` of the result of the redirect handling and `oneshot` from `main` to notify Hyper to shut down the server. This simplifies the flow a bit.